FileSystemWatcher is flaky

Very often it would lead to
```
Exception in thread "watcher" java.io.UncheckedIOException: java.nio.file.NoSuchFileException: /home/ilyak/w/c/Chronicle-Core/target/FileSystemWatcherTest-g5d8uj306y/dir2
	at java.nio.file.FileTreeIterator.fetchNextIfNeeded(FileTreeIterator.java:88)
	at java.nio.file.FileTreeIterator.hasNext(FileTreeIterator.java:104)
	at java.util.Iterator.forEachRemaining(Iterator.java:115)
	at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:482)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:472)
	at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:485)
	at net.openhft.chronicle.core.watcher.FileSystemWatcher.bootstrapPath(FileSystemWatcher.java:173)
	at net.openhft.chronicle.core.watcher.FileSystemWatcher.bootstrap(FileSystemWatcher.java:163)
	at net.openhft.chronicle.core.watcher.FileSystemWatcher.run(FileSystemWatcher.java:114)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.nio.file.NoSuchFileException: /home/ilyak/w/c/Chronicle-Core/target/FileSystemWatcherTest-g5d8uj306y/dir2
	at sun.nio.fs.UnixException.translateToIOException(UnixException.java:86)
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102)
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:107)
	at sun.nio.fs.UnixFileAttributeViews$Basic.readAttributes(UnixFileAttributeViews.java:55)
	at sun.nio.fs.UnixFileSystemProvider.readAttributes(UnixFileSystemProvider.java:144)
	at sun.nio.fs.LinuxFileSystemProvider.readAttributes(LinuxFileSystemProvider.java:99)
	at java.nio.file.Files.readAttributes(Files.java:1737)
	at java.nio.file.FileTreeWalker.getAttributes(FileTreeWalker.java:225)
	at java.nio.file.FileTreeWalker.visit(FileTreeWalker.java:276)
	at java.nio.file.FileTreeWalker.next(FileTreeWalker.java:372)
	at java.nio.file.FileTreeIterator.fetchNextIfNeeded(FileTreeIterator.java:95)
	... 13 more

org.junit.ComparisonFailure: 
```
walk() [is not safe to use](https://bugs.openjdk.java.net/browse/JDK-8039910), it seems the only safe alternative is to move over to walkFileTree()

--- a/src/main/java/net/openhft/chronicle/core/watcher/FileSystemWatcher.java
+++ b/src/main/java/net/openhft/chronicle/core/watcher/FileSystemWatcher.java
@@ -23,6 +23,7 @@ import net.openhft.chronicle.core.io.Closeable;
 
 import java.io.IOException;
 import java.nio.file.*;
+import java.nio.file.attribute.BasicFileAttributes;
 import java.util.*;
 import java.util.concurrent.BlockingQueue;
 import java.util.concurrent.ConcurrentHashMap;
@@ -59,16 +60,41 @@ public class FileSystemWatcher {
         Path base0 = Paths.get(base);
         Path base2 = base0.resolve(relative);
         if (Files.isDirectory(base2)) {
-            try (Stream<Path> paths = Files.walk(base2, 8, FileVisitOption.FOLLOW_LINKS)) {
-                paths.forEach(full -> addPath0(base0, full));
+            try {
+                Files.walkFileTree(base2, Collections.singleton(FileVisitOption.FOLLOW_LINKS), 8, new FileVisitor<Path>() {
+                    @Override
+                    public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs) {
+                        return visitFile(dir, attrs);
+                    }
 
+                    @Override
+                    public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) {
+                        addPath0(base0, file);
+
+                        return FileVisitResult.CONTINUE;
+                    }
+
+                    @Override
+                    public FileVisitResult visitFileFailed(Path file, IOException exc) {
+                        warnOnException(exc, base);
+
+                        return FileVisitResult.CONTINUE;
+                    }
+
+                    @Override
+                    public FileVisitResult postVisitDirectory(Path dir, IOException exc) {
+                        warnOnException(exc, base);
+
+                        return FileVisitResult.CONTINUE;
+                    }
+                });
             } catch (IOException e) {
-                Jvm.warn().on(FileSystemWatcher.class, "Couldn't walk path " + base, e);
+                warnOnException(e, base);
             }
             try {
                 bootstrapPath(listeners, base, relative);
             } catch (IOException e) {
-                Jvm.warn().on(FileSystemWatcher.class, "Couldn't walk path " + base, e);
+                warnOnException(e, base);
             }
         }
     }
@@ -169,8 +195,15 @@ public class FileSystemWatcher {
 
     private void bootstrapPath(List<WatcherListener> list, String base, String relative) throws IOException {
         Path full = Paths.get(base).resolve(relative);
-        try (Stream<Path> walk = Files.walk(full, 8, FileVisitOption.FOLLOW_LINKS)) {
-            walk.forEach(p -> {
+
+        Files.walkFileTree(full, Collections.singleton(FileVisitOption.FOLLOW_LINKS), 8, new FileVisitor<Path>() {
+            @Override
+            public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs) throws IOException {
+                return visitFile(dir, attrs);
+            }
+
+            @Override
+            public FileVisitResult visitFile(Path p, BasicFileAttributes attrs) throws IOException {
                 for (Iterator<WatcherListener> iterator = list.iterator(); iterator.hasNext(); ) {
                     WatcherListener listener = iterator.next();
                     String pToString = p.toString();
@@ -183,8 +216,23 @@ public class FileSystemWatcher {
                         iterator.remove();
                     }
                 }
-            });
-        }
+                return FileVisitResult.CONTINUE;
+            }
+
+            @Override
+            public FileVisitResult visitFileFailed(Path file, IOException exc) throws IOException {
+                warnOnException(exc, base);
+
+                return FileVisitResult.CONTINUE;
+            }
+
+            @Override
+            public FileVisitResult postVisitDirectory(Path dir, IOException exc) throws IOException {
+                warnOnException(exc, base);
+
+                return FileVisitResult.CONTINUE;
+            }
+        });
     }
 
     public void start() {
@@ -200,6 +248,12 @@ public class FileSystemWatcher {
             Thread.currentThread().interrupt();
         }
         Closeable.closeQuietly(watchService);
+    }
+
+    private void warnOnException(IOException exc, String base) {
+        if (exc != null) {
+            Jvm.warn().on(FileSystemWatcher.class, "Couldn't walk path " + base, exc);
+        }
     }
 
     static class PathInfo {

